src/Hazelcast.Net/Sql/SqlService.cs

```diff
@@ -112,36 +124,88 @@ public Task<long> ExecuteCommandAsync(string sql, SqlStatementOptions options =
+        /// </summary>
+        /// <param name="sql">Sql Statement.</param>
+        /// <param name="serializedParameters">Serialized parameters of the statement.</param>
+        /// <returns>Partition Id if successful, otherwise -1</returns>
+        private int GetPartitionIdOfQuery(string sql, List<IData> serializedParameters)
+        {
+            var partitionId = -1;
+
+            // Todo: Update condition after TPC implementation.
+            if (!_options.Networking.SmartRouting || !_queryPartitionArgumentCache.TryGetValue(sql, out var argIndex)) return partitionId;
+
+            if (argIndex >= 0 && argIndex < serializedParameters.Count)
+            {
+                var serializedKey = serializedParameters[argIndex];
+                partitionId = _cluster.State.Partitioner.GetPartitionId(serializedKey.PartitionHash);
+            }
+            else
+            {
+                _logger.IfDebug().LogDebug("Argument Index ({ArgIndex}) for query is out of range in parameters. SQL statement will send over a random connection.", argIndex);
+            }
+
+            return partitionId;
         }
 
-        private async Task<(SqlRowMetadata rowMetadata, SqlPage page)> FetchFirstPageAsync(SqlQueryId queryId, string sql, object[] parameters, SqlStatementOptions options, CancellationToken cancellationToken)
+        private async Task<(SqlRowMetadata rowMetadata, SqlPage page, int)> FetchFirstPageAsync(SqlQueryId queryId, string sql, object[] parameters, SqlStatementOptions options, CancellationToken cancellationToken)
         {
-            var result = await FetchAndValidateResponseAsync(queryId, sql, parameters, options, SqlResultType.Rows, cancellationToken).CfAwait();
+            var (result, partitionId) = await FetchAndValidateResponseAsync(queryId, sql, parameters, options, SqlResultType.Rows, cancellationToken).CfAwait();
             if (result.RowMetadata == null)
                 throw new HazelcastSqlException(_cluster.ClientId, SqlErrorCode.Generic, "Expected row set in the response but got update count.");
 
-            return (new SqlRowMetadata(result.RowMetadata), result.RowPage);
+            return (new SqlRowMetadata(result.RowMetadata), result.RowPage, partitionId);
         }
 
-        private async Task<SqlPage> FetchNextPageAsync(SqlQueryId queryId, int cursorBufferSize, CancellationToken cancellationToken)
+        private async Task<SqlPage> FetchNextPageAsync(SqlQueryId queryId, int cursorBufferSize, int partitionId, CancellationToken cancellationToken)
         {
             var requestMessage = SqlFetchCodec.EncodeRequest(queryId, cursorBufferSize);
-            var responseMessage = await _cluster.Messaging.SendAsync(requestMessage, cancellationToken).CfAwait();
+            var responseMessage = partitionId == -1 ?
+                await _cluster.Messaging.SendAsync(requestMessage, cancellationToken)
```

Comment:
can you put the `?` and `:` aligned:
```var ... = ...
    ? .....
    : .... ;```
as it's easier to read?
and then, you're issing a `CfAwait()` on one of the two calls

Reply:
don't think we understand each other, my idea was:
```
var responseMessage = partitionId == -1 
    ? await _cluster.Messaging.SendAsync(requestMessage, cancellationToken).CfAwait()
    : await _cluster.Messaging.SendToPartitionOwnerAsync(requestMessage, partitionId, cancellationToken).CfAwait();
```

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/798/commits/a91d960491af90459c71f0821c8ccfdd5e85d119